rust/src/storage/s3.rs

```diff
@@ -0,0 +1,687 @@
+//! AWS S3 storage backend. It only supports a single writer and is not multi-writer safe.
+
+use crate::builder::{s3_storage_options, str_option};
+use bytes::Bytes;
+use dynamodb_lock::{DynamoError, LockClient, LockItem, DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS};
+use futures::stream::BoxStream;
+use object_store::path::Path;
+use object_store::{
+    DynObjectStore, Error as ObjectStoreError, GetResult, ListResult, MultipartId, ObjectMeta,
+    ObjectStore, Result as ObjectStoreResult,
+};
+use rusoto_core::{HttpClient, Region};
+use rusoto_credential::AutoRefreshingProvider;
+use rusoto_sts::WebIdentityProvider;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::ops::Range;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::io::AsyncWrite;
+
+const STORE_NAME: &str = "DeltaS3ObjectStore";
+
+/// Error raised by storage lock client
+#[derive(thiserror::Error, Debug)]
+enum S3LockError {
```

Comment:
The basic idea of using private error types internally, but exposing one global error type to consumers resonates with me. This is more or less adopted from object store. Eventually we may want to apply this throughout delta-rs, with the aim to only expose `DetaTableError` publicly.